File: backend/src/main/scala/domains/usergroup/api/ListUserGroupSlugsForMember.scala
package domains.usergroup.api

import cats.effect.IO
import domains.auth.api.InternalOnlyApi
import domains.user.objects.Username
import domains.usergroup.objects.response.UserGroupSlugListResponse
import domains.usergroup.table.user_group.UserGroupTable
import org.http4s.Method
import shared.api.ApiPath

import java.sql.Connection

/** 内部成员用户组查询 API，供访问控制按用户获取其加入的用户组 slug。 */
object ListUserGroupSlugsForMember extends InternalOnlyApi[Username, UserGroupSlugListResponse]:

  override val method: Method = Method.POST
  override val path: ApiPath = ApiPath("/api/internal/user-groups/member-slugs")

  /** 查询指定用户所属用户组 slug，并按字典序返回。 */
  override def plan(connection: Connection, username: Username): IO[UserGroupSlugListResponse] =
    UserGroupTable
      .listGroupSlugsForMember(connection, username)
      .map(slugs => UserGroupSlugListResponse(slugs.toList.sortBy(_.value)))

File: frontend/src/apis/usergroup/ListUserGroupSlugsForMember.ts
import type { APIMessage } from '@/system/api/api-message'
import type { UserGroupSlugListResponse } from '@/objects/usergroup/response/UserGroupSlugListResponse'
import type { Username } from '@/objects/user/Username'

/** 内部查询成员所属用户组 slug 的请求体；用户名通过 body 传递。 */
type ListUserGroupSlugsForMemberBody = {
  username: Username
}

/** 查询某用户所在用户组 slug 列表；用于权限策略选择器或内部授权辅助。 */
export class ListUserGroupSlugsForMember implements APIMessage<UserGroupSlugListResponse> {
  declare readonly responseType?: UserGroupSlugListResponse
  readonly method = 'POST'
  readonly apiPath = 'internal/user-groups/member-slugs'
  private readonly username: Username

  constructor(username: Username) {
    this.username = username
  }

  body(): ListUserGroupSlugsForMemberBody {
    return { username: this.username }
  }
}
